R/codecov.R

```diff
@@ -0,0 +1,142 @@
+    codecov_query <- list(branch = Sys.getenv("BRANCH_NAME"),
+                          build = Sys.getenv("SEMAPHORE_BUILD_NUMBER"),
+                          owner = strsplit(Sys.getenv("SEMAPHORE_REPO_SLUG"), "/")[0],
+                          repo = strsplit(Sys.getenv("SEMAPHORE_REPO_SLUG"), "/")[1],
+                          commit = Sys.getenv("REVISION"))
+  # --------
+  # drone.io
+  # --------
+  } else if (Sys.getenv("CI") == "true" && Sys.getenv("DRONE") == "true") {
+    # http://docs.drone.io/env.html
+    codecov_url <- paste0(base_url, "/upload/v2?service=drone.io") # nolint
+    codecov_query <- list(branch = Sys.getenv("DRONE_BRANCH"),
+                          build = Sys.getenv("DRONE_BUILD_NUMBER"),
+                          build_url = Sys.getenv("DRONE_BUILD_URL"),
+                          commit = Sys.getenv("DRONE_COMMIT"))
+  # --------
+  # AppVeyor
+  # --------
+  } else if (Sys.getenv("CI") == "True" && Sys.getenv("APPVEYOR") == "True") {
+    # http://www.appveyor.com/docs/environment-variables
+    codecov_url <- paste0(base_url, "/upload/v2?service=AppVeyor") # nolint
+    codecov_query <- list(branch = Sys.getenv("APPVEYOR_REPO_BRANCH"),
+                          build = Sys.getenv("APPVEYOR_BUILD_NUMBER"),
+                          owner = strsplit(Sys.getenv("APPVEYOR_REPO_NAME"), "/")[0],
+                          repo = strsplit(Sys.getenv("APPVEYOR_REPO_NAME"), "/")[1],
+                          commit = Sys.getenv("APPVEYOR_REPO_COMMIT"))
+  # -------
+  # Wercker
+  # -------
+  } else if (Sys.getenv("CI") == "true" && Sys.getenv("WERCKER_GIT_BRANCH")) {
+    # http://devcenter.wercker.com/articles/steps/variables.html
+    codecov_url <- paste0(base_url, "/upload/v2?service=werker") # nolint
+    codecov_query <- list(branch = Sys.getenv("WERCKER_GIT_BRANCH"),
+                          build = Sys.getenv("WERCKER_MAIN_PIPELINE_STARTED"),
+                          owner = Sys.getenv("WERCKER_GIT_OWNER"),
+                          repo = Sys.getenv("WERCKER_GIT_REPOSITORY"),
+                          commit = Sys.getenv("WERCKER_GIT_COMMIT"))
+  # ---------
+  # Local GIT
+  # ---------
+  } else {
+    branch <- local_branch()
+    codecov_url <- paste0(base_url, "/upload/v2") # nolint
+    codecov_query <- list(branch = ifelse(branch == "HEAD", "master", branch),
+                          commit = trim(system("git rev-parse HEAD", intern = TRUE)))
+  }
+
+  if (Sys.getenv("CODECOV_TOKEN") != "") {
+    codecov_query["token"] <- Sys.getenv("CODECOV_TOKEN")
+  }
+
+  coverage <- to_codecov(package_coverage(path, relative_path = TRUE, ...))
+
+  name <- tempfile()
+  con <- file(name)
+  writeChar(con = con, coverage, eos = NULL)
+  close(con)
+  on.exit(unlink(name))
+  httr::content(httr::POST(url = codecov_url, query = codecov_query, body = list(json_file = httr::upload_file(name))))
+}
+
+to_codecov <- function(x) {
+  coverages <- per_line(x)
```

Comment:
It took a while to debug this, but it seems that the Codecov API expects an extra `NULL` or `NA` compared to Coverall, so the returned list by `per_line` should be tweaked with a leading `NA`. Quick and dirty workaround for this line:

```r
coverages <- lapply(covr:::per_line(package_coverage()), function(x) c(NA, x))
```

Now the two services are in sync:
* https://coveralls.io/builds/2105425/source?filename=.%2FR%2Fr-test.R
* https://codecov.io/github/daroczig/r-test/R/r-test.R?ref=master